### PR TITLE
Adds Cyclical Learning Rate and Momentum

### LINKED ISSUE #1909: [Feature Request] Cyclical Learning Rates
[Cyclical learning rates](https://arxiv.org/abs/1506.01186) seem to be becoming more popular. Given that we now have a [lr scheduler](https://github.com/pytorch/pytorch/pull/1370), it would be nice to have this available as another schedule scheme. Pinging @ajbrock, who has experience with this (unlike me haha).

### PR DESCRIPTION
**Pull request by sampepose**
This implements a cyclical learning rate (CLR) schedule with an optional inverse cyclical momentum. More info about CLR: https://github.com/bckenstler/CLR

This is finishing what #2016 started. Resolves #1909.

### CODE REVIEW

**Comment by soumith** [comment]
thanks Sam (and @thomasjpfan too). Let me know if you need a review / re-review.

**Comment by sampepose** [comment]
Hey @soumith! I'd appreciate a review, thank you!

**Comment by thomasjpfan** [comment]
It would be good if this PR or #2016 gets merged. (I started #2016 in 2017) 😄

**Comment by sampepose** [comment]
ping @soumith :)

File: torch/optim/lr_scheduler.py [hunk 1]
```diff
@@ -427,3 +428,210 @@ def state_dict(self):
...
+        A cycle that scales initial amplitude by gamma**(cycle iterations) at each
+        cycle iteration.
+
+    This implementation was adapted from the github repo: `bckenstler/CLR`_
+
+    Args:
+        optimizer (Optimizer): Wrapped optimizer.
+        base_lr (float or list): Initial learning rate which is the
+            lower boundary in the cycle for each parameter group.
+            Default: 0.001
```
Comment (soumith) [inline torch/optim/lr_scheduler.py]: I'd suggest to not have a default for the base learning rate. It makes it too easy to shoot oneself in the foot

File: torch/optim/lr_scheduler.py [hunk 2]
```diff
@@ -427,3 +428,210 @@ def state_dict(self):
...
+        max_lr (float or list): Upper learning rate boundaries in the cycle
+            for each parameter group. Functionally,
+            it defines the cycle amplitude (max_lr - base_lr).
+            The lr at any cycle is the sum of base_lr
+            and some scaling of the amplitude; therefore
+            max_lr may not actually be reached depending on
+            scaling function. Default: 0.006
```
Comment (soumith) [inline torch/optim/lr_scheduler.py]: You either want to make the max_lr default be a multiple of base_lr, or simply remove a default

File: torch/optim/lr_scheduler.py [hunk 3]
```diff
@@ -427,3 +428,210 @@ def state_dict(self):
...
+            Values correspond to policies detailed above.
+            If scale_fn is not None, this argument is ignored.
+            Default: 'triangular'
+        gamma (float): Constant in 'exp_range' scaling function:
+            gamma**(cycle iterations)
+            Default: 1.0
+        scale_fn (function): Custom scaling policy defined by a single
+            argument lambda function, where
+            0 <= scale_fn(x) <= 1 for all x >= 0.
+            mode parameter is ignored
```
Comment (soumith) [inline torch/optim/lr_scheduler.py]: " mode parameter is ignored" -> "If specified, then `mode` is ignored"

File: torch/optim/lr_scheduler.py [hunk 4]
```diff
@@ -427,3 +428,210 @@ def state_dict(self):
...
+            lower boundary in the cycle for each parameter group.
+            Default: 0.8
+        max_momentum (float or list): Upper momentum boundaries in the cycle
+            for each parameter group. Functionally,
+            it defines the cycle amplitude (max_momentum - base_momentum).
+            The momentum at any cycle is the difference of max_momentum
+            and some scaling of the amplitude; therefore
+            base_momentum may not actually be reached depending on
+            scaling function. Default: 0.9
+        last_batch_idx (int): The index of the last batch. Default: -1
```
Comment (soumith) [inline torch/optim/lr_scheduler.py]: what exactly does `-1` do here?

Comment (soumith) [reply torch/optim/lr_scheduler.py]: also, it's not clear what this is used for, explaining what the index of the last batch is, or where / why it is used will help

File: torch/optim/lr_scheduler.py [hunk 5]
```diff
@@ -427,3 +428,210 @@ def state_dict(self):
...
+
+    Example:
+        >>> optimizer = torch.optim.SGD(model.parameters(), lr=0.1, momentum=0.9)
+        >>> scheduler = torch.optim.CyclicLR(optimizer)
+        >>> data_loader = torch.utils.data.DataLoader(...)
+        >>> for epoch in range(10):
+        >>>     for batch in data_loader:
+        >>>         scheduler.step()
```
Comment (soumith) [inline torch/optim/lr_scheduler.py]: Shouldn't it be:

```
train_batch(...)
scheduler.step()
```

File: torch/optim/lr_scheduler.py [hunk 6]
```diff
@@ -427,3 +428,210 @@ def state_dict(self):
...
+        >>>         train_batch(...)
+
+    .. _Cyclical Learning Rates for Training Neural Networks: https://arxiv.org/abs/1506.01186
+    .. _bckenstler/CLR: https://github.com/bckenstler/CLR
+    """
+
+    def __init__(self,
+                 optimizer,
+                 base_lr=1e-3,
```
Comment (soumith) [inline torch/optim/lr_scheduler.py]: comments above about default values

File: torch/optim/lr_scheduler.py [hunk 7]
```diff
@@ -427,3 +428,210 @@ def state_dict(self):
...
+        self.optimizer = optimizer
+
+        base_lrs = self._format_param('base_lr', optimizer, base_lr)
+        if last_batch_idx == -1:
+            for lr, group in zip(base_lrs, optimizer.param_groups):
+                group['lr'] = lr
+
+        self.max_lrs = self._format_param('max_lr', optimizer, max_lr)
+
+        step_size_down = step_size_down or step_size_up
```
Comment (soumith) [inline torch/optim/lr_scheduler.py]: do you mean `step size_down = step_size_down if step_size_down is not None else step_size_up`

File: torch/optim/lr_scheduler.py [hunk 8]
```diff
@@ -427,3 +428,210 @@ def state_dict(self):
...
+        self.total_size = float(step_size_up + step_size_down)
```
Comment (soumith) [inline torch/optim/lr_scheduler.py]: it seems safer to actually convert step_size_up and step_size_down to float immediately on entry into the function.

**Review by soumith — CHANGES_REQUESTED** [review]
comments are minor, but need necessary changes.

Thanks, the PR looks pretty good.

**Review by soumith — APPROVED** [review]
lgtm. thank you!
